### PR TITLE
Support in place transform

### LINKED ISSUE #61: Support in-place transform

### PR DESCRIPTION
**Pull request by yasahi-hpc**
Resolves #61

This PR aims at supporting in-place transform.
Following changes are made.

- [x] Add a function `are_aliasing` to check if the input and output views are alias
- [x] Add a test function for `are_aliasing`
- [x] Improve `get_extents` function to compute extents for `in-place` case
- [x] Use `in-place` options for `ROCM` and `SYCL` in case of `in-place` transforms
- [x] Add tests for 1D and 2D `in-place` transforms

### CODE REVIEW

File: fft/unit_test/Test_Transform.cpp
Comment (tpadioleau) [inline fft/unit_test/Test_Transform.cpp:1705]:
```suggestion
  auto sub_xr_padded =
      Kokkos::subview(xr_padded, Kokkos::ALL, Kokkos::pair<int, int>(0, n1));
```

File: fft/unit_test/Test_Transform.cpp
Comment (tpadioleau) [inline fft/unit_test/Test_Transform.cpp:1759]:
```suggestion
        inv_xr_hat_padded, Kokkos::ALL, Kokkos::pair<int, int>(0, n1));
```

File: fft/unit_test/Test_Utils.hpp
Comment (tpadioleau) [inline fft/unit_test/Test_Utils.hpp:148]:
```suggestion
              auto phase = 2 * I * Kokkos::numbers::pi * static_cast<real_value_type>(i) /
```

File: fft/unit_test/Test_Utils.hpp
Comment (tpadioleau) [inline fft/unit_test/Test_Utils.hpp:102]:
```suggestion
              auto phase = -2 * I * Kokkos::numbers::pi * static_cast<real_value_type>(i) /
```

If you support recent enough Kokkos ?

File: common/unit_test/Test_Helpers.cpp
Comment (tpadioleau) [inline common/unit_test/Test_Helpers.cpp:390]: Could you add https://clang.llvm.org/docs/ClangFormatStyleOptions.html#insertnewlineateof to .clang-format ?

File: fft/unit_test/Test_Transform.cpp
Comment (tpadioleau) [inline fft/unit_test/Test_Transform.cpp:1683]: Same I would only keep `RealView2DType` and `ComplexView2DType`

**Comment by yasahi-hpc** [comment]
> Can you confirm that fftw, cufft and hipfft internally do a check for an in-place execution ? No need to modify anything on our side ?

Thank you for reviews

I expect that `fftw3` and `cufft` are doing checks internally, but it is not very clear from documentations. 
If I understand, all we need is just passing same pointers as input and output
`hipfft` in turn checks internally (see https://github.com/ROCm/hipFFT/blob/f9a5de3d5bfaa08b52d964bf6501478ee71398f3/clients/hipfft_params.h#L512-L516)

File: fft/unit_test/Test_Transform.cpp
Comment (yasahi-hpc) [reply fft/unit_test/Test_Transform.cpp:115]: Actually, `allclose` should take an execution space. In this PR, I will add a fence as a temporary solution

File: common/unit_test/Test_Helpers.cpp
Comment (yasahi-hpc) [reply common/unit_test/Test_Helpers.cpp:390]: Can I do this first in another PR to make this PR simpler?

File: common/src/KokkosFFT_utils.hpp
Comment (yasahi-hpc) [reply common/src/KokkosFFT_utils.hpp:23]: Sure

File: fft/src/KokkosFFT_Plans.hpp
Comment (yasahi-hpc) [reply fft/src/KokkosFFT_Plans.hpp:368]: Sure

File: common/unit_test/Test_Utils.cpp
Comment (yasahi-hpc) [reply common/unit_test/Test_Utils.cpp:573]: Sure

File: fft/unit_test/Test_Utils.hpp
Comment (yasahi-hpc) [reply fft/unit_test/Test_Utils.hpp:102]: Sure

File: common/unit_test/Test_Helpers.cpp
Comment (tpadioleau) [reply common/unit_test/Test_Helpers.cpp:390]: sure

File: fft/unit_test/Test_Transform.cpp
Comment (tpadioleau) [reply fft/unit_test/Test_Transform.cpp:115]: works for me

**Comment by tpadioleau** [comment]
> I expect that `fftw3` and `cufft` are doing checks internally, but it is not very clear from documentations. If I understand, all we need is just passing same pointers as input and output `hipfft` in turn checks internally (see https://github.com/ROCm/hipFFT/blob/f9a5de3d5bfaa08b52d964bf6501478ee71398f3/clients/hipfft_params.h#L512-L516)

That is also my understanding reading cufft.

File: common/src/KokkosFFT_layouts.hpp
Comment (yasahi-hpc) [reply common/src/KokkosFFT_layouts.hpp:25]: Can I do that in another PR? 
I need to modify the tests for this function as well

File: fft/unit_test/Test_Utils.hpp
Comment (yasahi-hpc) [reply fft/unit_test/Test_Utils.hpp]: I got a compile error for nvcc

`/work/fft/unit_test/Test_Utils.hpp(102): error: identifier "Kokkos::numbers::pi" is undefined in device code`

Have you already succeeded to use this in DDC?

**Comment by yasahi-hpc** [comment]
@tpadioleau 
May I have your second review please

**Comment by yasahi-hpc** [comment]
@tpadioleau 
It seems that we need to keep pi in double precision.

**Comment by tpadioleau** [comment]
> @tpadioleau It seems that we need to keep pi in double precision.

Not precise enough otherwise ?

**Comment by yasahi-hpc** [comment]
@tpadioleau Seems OK now.
If you feel good, I will merge this

File: fft/unit_test/Test_Utils.hpp
Comment (tpadioleau) [inline fft/unit_test/Test_Utils.hpp:81]:
```suggestion
void fft1(const ViewType& in, const ViewType& out) {
```

File: fft/unit_test/Test_Utils.hpp
Comment (tpadioleau) [inline fft/unit_test/Test_Utils.hpp:129]:
```suggestion
void ifft1(const ViewType& in, const ViewType& out) {
```

File: fft/unit_test/Test_Utils.hpp
Comment (tpadioleau) [inline fft/unit_test/Test_Utils.hpp:137]:
```suggestion

  static_assert(KokkosFFT::Impl::is_complex_v<value_type>,
                "ifft1: ViewType must be complex");

  constexpr double pi = Kokkos::numbers::pi;
```

File: fft/unit_test/Test_Utils.hpp
Comment (tpadioleau) [inline fft/unit_test/Test_Utils.hpp:89]:
```suggestion

  static_assert(KokkosFFT::Impl::is_complex_v<value_type>,
                "fft1: ViewType must be complex");

  constexpr double pi = Kokkos::numbers::pi;
```

**Review by tpadioleau — DISMISSED** [review]
Seems ok to me

File: fft/unit_test/Test_Utils.hpp
Comment (yasahi-hpc) [reply fft/unit_test/Test_Utils.hpp:137]: Why you think is better?
According to their usage, it would be
`constexpr auto pi = Kokkos::numbers::pi_v<double>; ?`

https://github.com/kokkos/kokkos/blob/e6e5c4598d16d756db62225ab7f937ee833bd660/algorithms/src/Kokkos_Random.hpp#L855

**Comment by yasahi-hpc** [comment]
@tpadioleau Thank you for review. I think I have fixed accordingly

File: fft/unit_test/Test_Utils.hpp
Comment (tpadioleau) [reply fft/unit_test/Test_Utils.hpp]: `pi` is the `double` version of `pi_v`, just a shortcut when you usually need `double` precision

**Review by tpadioleau — APPROVED** [review]
ok for me

**Comment by yasahi-hpc** [comment]
I will merge this. Thank you again for your review

**Pull request merged**
